README

DESCRIPTION

	hping6 is a network tool able to send custom ICMP/UDP/TCP
	packets and to display target replies like ping do with
	ICMP replies. hping6 handle fragmentation, arbitrary packet
	body and size and can be used in order to transfer files
	under supported protocols. Using hping6 you are able at
	least to perform the following jobs:

	- Test firewall rules
	- [spoofed] port scanning
	- Test net performance using differents protocols,
	  packet size, TOS (type of service) and fragmentation.
	- Path MTU discovery
	- Files trasfering even between really fascist firewall rules.
	- Traceroute like under different protocols.
	- Firewalk like usage.
	- Remote OS fingerprint.
	- TCP/IP stack auditing.
	- IPv6 support
	- A lot of others.

	It's also really a good didactic tool to learn TCP/IP.

	hping6 is developed and manteined by antirez@invece.org
	and comes under GPL version 2 of license. Developing is
	open so you can send me patches/suggestions/affronts without
	inhibitions.

	Thanks to the following peoples for them suggestions/tests/tools/code:

	Valeriano Bedeschi	<vale@seclab.com>
	Lorenzo Cavallaro	<sullivan@seclab.com>
	awgn roofing		<root@roof.penguinpowered.com>
	Darren Reed		<avalon@COOMBS.ANU.EDU.AU>
	Alfonso De Gregorio	<dira@speedcom.it>
	Mika			<mika@qualys.com>
	Yann Berthier		<Yann.Berthier@hsc.fr>
	Lance Spitzner		<lance@spitzner.net>
	Stephane Aubert		<Stephane.Aubert@hsc.fr>
	Nicolas Jombart		<Nicolas.Jombart@hsc.fr>
	Denis Ducamp		<Denis.Ducamp@hsc.fr>
	Francesco Potorti`	<pot@gnu.org>
	Stefano Brandimarte	<stevens@alicom.com>
	"roy kozzer"		<royk50@hotmail.com>
	Jason Lunz		<j@trellisinc.com>
	Domenico Andreoli	<cavok@filibusta.crema.unimi.it>
	Gian-Luca Dei Rossi	<acaso@venezia.linux.it>
	Marco D'Itri		<md@linux.it>
	Rui Miguel Barbosa Machado <rmbm@rccn.net>
	Daniel Ginsburg		<dbg@nm.ru>
	Steve Bleazard		<steve@bleazard.com>
	David Coppa		<coffeec@tin.it>

	Many other, I don't remember.

	Also vim developers, ee.lbl.gov for tcpdump and GNU in general.

	IPv6 support, flood modes:
	Matyas Koszik		<koszik@atw.hu>

REQUIREMENTS

	A supported unix-like OS, gcc, root access.

	see PORTINGS file for supported systems list.

USAGE

	See the man page and HPING6-HOWTO.

INSTALLATION

	see INSTALL file.

have fun,
antirez


NEWS
This short document is for users of hping-beta54 or prior versions
and helps to exploit all the new features of this hping6 release in a
short time. You may want to read the new man page anyway but the
following will help for sure:

=== release candidate 3 news

In this release a nasty bug with the checksum code was fixed.
If you experimented strange problems like some kind of packet
generated with a wrong checksum try this version.

Try the --scan option in the command line to see the port-scanner features.

  Example of the --scan option usage:

# hping3 --scan known 1.2.3.4

Scanning 1.2.3.4 (1.2.3.4), port known
245 ports to scan, use -V to see all the replies
+----+-----------+---------+---+-----+-----+-----+
|port| serv name |  flags  |ttl| id  | win | len |
+----+-----------+---------+---+-----+-----+-----+
    9 discard    : .S..A...  64     0 32767    44
   13 daytime    : .S..A...  64     0 32767    44
   21 ftp        : .S..A...  64     0 32767    44
   22 ssh        : .S..A...  64     0 32767    44
   25 smtp       : .S..A...  64     0 32767    44
   37 time       : .S..A...  64     0 32767    44
   80 www        : .S..A...  64     0 32767    44
  111 sunrpc     : .S..A...  64     0 32767    44
  113 auth       : .S..A...  64     0 32767    44
  631 ipp        : .S..A...  64     0 32767    44
 3306 mysql      : .S..A...  64     0 32767    44
 6000 x11        : .S..A...  64     0 32767    44
 6667 ircd       : .S..A...  64     0  3072    44
All replies received. Done.
Not responding ports: 

Check the man page for more information on the scan mode.

=== release candidate 2 news

. Now hping is able to send/parse source routed IP headers.
  See the manpage for more info.

. Hping was almost rewrote, at least all the most important parts.
  You should experiment a more readable, compact, fast to compile
  code.

. The new option parsing code allows you to specify abbreviated
  options. you can now use for example --tcp-ti instead of --tcp-timestamp
  and so on.

. The new feature rand-dest allows to send the packet to random
  IP addresses. This is very useful to do some Internet survey
  or large subnet random scanning.

  For example the follow command line will send TCP packets with the
  SYN flag on to the port 80 of the 192.168.0.0/16 address space:

  hping 192.168.x.x --rand-dest -p 80 -S

  Every occurrence of 'x' is substituted with a random number
  in the 0-255 range.

. The new feature rand-source allows to send packets with random
  source addresses. Useful to test some DoS condition against firewalls
  or TCP/IP stacks that implements some per-IP basis information
  recording.

. The output was enhanced and fixed a bit.

. The "force incremental dest port" option (++<port>) now works with UDP
  packets and works better with TCP, since it is more selective
  with the incoming responses.

. Now you should be really able to set the sequence and acknowledge
  number of the TCP packets. The rc1 code was broken because
  atoi() was used to get a long unsigned value.

. The documentation (and the french translation) was updated
  to reflect the changes.

=== release candidate 1 news

. Now hping works better on BSD, and works on Solaris. It should
  be many times simplest to port it to an unsupported platform.
  Problems with systems that uses 32bit pids are now fixed.

. The output is different to be more parseable and compact, example:

  len=46 ip=192.168.1.1 flags=RA DF seq=0 ttl=255 id=0 win=0 rtt=0.5 ms

  now the presence of the Don't fragment IP flag is signaled with 'DF'.
  all the fields with a value are in the form 'field=value'.

. To specify the outgoing interface with -I is no longer needed,
  hping will try to detect the right interface according to the
  system routing table. Of course you can override it using -I.

. Instead to specify -i u10000 to get a speed of ten packets for second
  you can just use --fast.

. Now --traceroute (-T) implies --ttl 1. You can override this using --ttl.

. Using hping as traceroute you have now RTT informations about the
  hops.

. You can monitor a specific hop in traceroute mode, using the following
  syntax:

  hping6 -T www.yahoo.com --tr-keep-ttl --ttl 5

  see the output:

  HPING www.yahoo.com (ippp0 64.58.76.177): NO FLAGS are set, 40 headers + 0 dat
  a bytes
  5->TTL 0 during transit from 144.232.234.57  (sl-gw18-nyc-2-2.sprintlink.net)
  5->RTT was: 136.9 ms
  5->TTL 0 during transit from 144.232.234.57  (sl-gw18-nyc-2-2.sprintlink.net)
  5->RTT was: 136.8 ms
  5->TTL 0 during transit from 144.232.234.57  (sl-gw18-nyc-2-2.sprintlink.net)
  5->RTT was: 136.9 ms
  5->TTL 0 during transit from 144.232.234.57  (sl-gw18-nyc-2-2.sprintlink.net)
  5->RTT was: 136.7 ms

 --- www.yahoo.com hping statistic ---
 4 packets tramitted, 0 packets received, 100% packet loss
 round-trip min/avg/max = 136.7/136.8/136.9 ms

 you get only information about the 5 hop, after Ctrl+C the round-trip
 min/avg/max is calculated using the rtt of this hop.

. Using the option --tr-stop you can obtain that hping will exit
  when the first matching packet that isn't an ICMP time exceeded
  in transit is received, like the original traceroute. Without
  this hping continue to send packets to the target host forever.

. You can use --tr-no-rtt to suppress the rtt information in traceroute
  mode.

. With the --tcp-timestamp feature you can guess the uptime of some
  remote systems. Example:

HPING www.hping.org (ippp0 192.70.106.166): S set, 40 headers + 0 data bytes
56 bytes from 192.70.106.166: flags=SA seq=0 ttl=49 id=28881 win=16080 rtt=105.0 ms
  TCP timestamp: 258597761

56 bytes from 192.70.106.166: flags=SA seq=1 ttl=49 id=28882 win=16080 rtt=105.4 ms
  TCP timestamp: 258597860
  HZ seems 100
  System uptime seems: 29 days, 22 hours, 19 minutes, 38 seconds

56 bytes from 192.70.106.166: flags=SA seq=2 ttl=49 id=28883 win=16080 rtt=105.1 ms
  TCP timestamp: 258597960
  HZ seems 100

usage.c
/* 
 * $smu-mark$ 
 * $name: usage.c$ 
 * $author: Salvatore Sanfilippo <antirez@invece.org>$ 
 * $copyright: Copyright (C) 1999 by Salvatore Sanfilippo$ 
 * $license: This software is under GPL version 2 of license$ 
 * $date: Fri Nov  5 11:55:50 MET 1999$ 
 * $rev: 8$ 
 */ 

/* $Id: usage.c,v 1.13 2003/08/07 23:55:55 antirez Exp $ */

#include <stdlib.h>
#include <stdio.h>
#include <stdlib.h>

void	show_usage(void)
{
	printf(
"usage: hping host [options]\n"
"  -h  --help      show this help\n"
"  -v  --version   show version\n"
"  -c  --count     packet count\n"
"  -i  --interval  wait (uX for X microseconds, for example -i u1000)\n"
"      --fast      alias for -i u10000 (10 packets for second)\n"
"      --flood	   sent packets as fast as possible. Don't show replies.\n"
"  -n  --numeric   numeric output\n"
"  -q  --quiet     quiet\n"
"  -I  --interface interface name (otherwise default routing interface)\n"
"  -V  --verbose   verbose mode\n"
"  -D  --debug     debugging info\n"
"  -z  --bind      bind ctrl+z to ttl           (default to dst port)\n"
"  -Z  --unbind    unbind ctrl+z\n"
"Mode\n"
"  default mode     TCP\n"
"  -0  --rawip      RAW IP mode\n"
"  -1  --icmp       ICMP mode\n"
"  -2  --udp        UDP mode\n"
"  -6  --ipv6       IPv6 mode\n"
"  -8  --scan       SCAN mode.\n"
"                   Example: hping --scan 1-30,70-90 -S www.target.host\n"
"  -9  --listen     listen mode\n"
"IP\n"
"  -a  --spoof      spoof source address\n"
"  --rand-dest      random destionation address mode. see the man.\n"
"  --rand-source    random source address mode. see the man.\n"
"  -t  --ttl        ttl (default 64)\n"
"  -N  --id         id (default random)\n"
"  -W  --winid      use win* id byte ordering\n"
"  -r  --rel        relativize id field          (to estimate host traffic)\n"
"  -f  --frag       split packets in more frag.  (may pass weak acl)\n"
"  -x  --morefrag   set more fragments flag\n"
"  -y  --dontfrag   set dont fragment flag\n"
"  -g  --fragoff    set the fragment offset\n"
"  -m  --mtu        set virtual mtu, implies --frag if packet size > mtu\n"
"  -o  --tos        type of service (default 0x00), try --tos help\n"
"  -G  --rroute     includes RECORD_ROUTE option and display the route buffer\n"
"  --lsrr           loose source routing and record route\n"
"  --ssrr           strict source routing and record route\n"
"  -H  --ipproto    set the IP protocol field, only in RAW IP mode\n"
"ICMP\n"
"  -C  --icmptype   icmp type (default echo request)\n"
"  -K  --icmpcode   icmp code (default 0)\n"
"      --force-icmp send all icmp types (default send only supported types)\n"
"      --icmp-gw    set gateway address for ICMP redirect (default 0.0.0.0)\n"
"      --icmp-ts    Alias for --icmp --icmptype 13 (ICMP timestamp)\n"
"      --icmp-addr  Alias for --icmp --icmptype 17 (ICMP address subnet mask)\n"
"      --icmp-help  display help for others icmp options\n"
"UDP/TCP\n"
"  -s  --baseport   base source port             (default random)\n"
"  -p  --destport   [+][+]<port> destination port(default 0) ctrl+z inc/dec\n"
"  -k  --keep       keep still source port\n"
"  -w  --win        winsize (default 64)\n"
"  -O  --tcpoff     set fake tcp data offset     (instead of tcphdrlen / 4)\n"
"  -Q  --seqnum     shows only tcp sequence number\n"
"  -b  --badcksum   (try to) send packets with a bad IP checksum\n"
"                   many systems will fix the IP checksum sending the packet\n"
"                   so you'll get bad UDP/TCP checksum instead.\n"
"  -M  --setseq     set TCP sequence number\n"
"  -L  --setack     set TCP ack\n"
"  -F  --fin        set FIN flag\n"
"  -S  --syn        set SYN flag\n"
"  -R  --rst        set RST flag\n"
"  -P  --push       set PUSH flag\n"
"  -A  --ack        set ACK flag\n"
"  -U  --urg        set URG flag\n"
"  -X  --xmas       set X unused flag (0x40)\n"
"  -Y  --ymas       set Y unused flag (0x80)\n"
"  --tcpexitcode    use last tcp->th_flags as exit code\n"
"  --tcp-timestamp  enable the TCP timestamp option to guess the HZ/uptime\n"
"Common\n"
"  -d  --data       data size                    (default is 0)\n"
"  -E  --file       data from file\n"
"  -e  --sign       add 'signature'\n"
"  -j  --dump       dump packets in hex\n"
"  -J  --print      dump printable characters\n"
"  -B  --safe       enable 'safe' protocol\n"
"  -u  --end        tell you when --file reached EOF and prevent rewind\n"
"  -T  --traceroute traceroute mode              (implies --bind and --ttl 1)\n"
"  --tr-stop        Exit when receive the first not ICMP in traceroute mode\n"
"  --tr-keep-ttl    Keep the source TTL fixed, useful to monitor just one hop\n"
"  --tr-no-rtt	    Don't calculate/show RTT information in traceroute mode\n"
"ARS packet description (new, unstable)\n"
"  --apd-send       Send the packet described with APD (see docs/APD.txt)\n"
	);
	exit(0);
};

void tos_help(void)
{
	printf(
"tos help:\n"
"          TOS Name                Hex Value           Typical Uses\n"
"\n"
"       Minimum Delay                 10               ftp, telnet\n"
"       Maximum Throughput            08               ftp-data\n"
"       Maximum Reliability           04               snmp\n"
"       Minimum Cost                  02               nntp\n"
	);
	exit(0);
}

void icmp_help(void)
{
	printf(
"ICMP help:\n"
" ICMP concerned packet options:\n"
"  --icmp-ipver     set ip version               ( default 4 )\n"
"  --icmp-iphlen    set ip header lenght         ( default IPHDR_SIZE >> 2)\n"
"  --icmp-iplen     set ip total lengtht         ( default real lenght )\n"
"  --icmp-ipid      set ip id                    ( default random )\n"
"  --icmp-ipproto   set ip protocol              ( default IPPROTO_TCP )\n"
"  --icmp-ipsrc     set ip source                ( default 0.0.0.0 )\n"
"  --icmp-ipdst     set ip destination           ( default 0.0.0.0 )\n"
"  --icmp-srcport   set tcp/udp source port      ( default random )\n"
"  --icmp-dstport   set tcp/udp destination port ( default random )\n"
"  --icmp-cksum     set icmp checksum            ( default the right cksum)\n"
	);
	exit(0);
}

void route_help(void)
{
    printf(
"route help:\n"
"	A route has the following format: [ptr:]IP1[/IP2[/IP3...]]\n"
"	where ptr is the exact value of the pointer that will be used for the IP\n"
"	option (be careful, no check is performed on this pointer), and defaults\n"
"	to 8, or 4 if provided route is too short for 8;\n"
"	and each IPx field is an IP address to include in the source route.\n");
}
